### PR TITLE
swift-format tool

### PR DESCRIPTION
**Pull request by danielmartin**
#### What's in this pull request?

This patch introduces a `swift-format` tool. This tool is designed to format Swift files or file ranges according to some parameters like amount of spaces, or tab width.

Apparently, this completes https://bugs.swift.org/browse/SR-146 Planned future work revolves around introducing an advanced formatting mechanism based on styles.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by gribozavr** [comment]
@danielmartin Thank you for your contribution!

Not a real review, but a few pieces of advice:

- Please avoid using `llvm::cl::opt` in user-facing tools.  Please use the same command-line parsing mechanism as the frontend and the driver use.

- Adding a new binary is very heavy-weight, everything needs to be linked again, and we would be shipping another copy of LLVM, Clang and Swift embedded in `swift-format`.  Could you model your tool after `tools/driver/autolink_extract_main.cpp`?

- What is the origin of the code submitted as tests?  Just wanted to confirm that you wrote that code and can contribute it under the license terms of the Swift project.

- Do we need that much code in tests?  It is unclear what particular features of the formatting engine you are testing.  It would be best to have tests be clearly focused on specific features.  Using large blobs of code for SourceKit tests has proven to be problematic, because the outputs change unpredictably as the compiler and the standard library evolve.  For example, take a look at the history of `test/SourceKit/DocSupport/doc_clang_module.swift.response` and `test/SourceKit/DocSupport/doc_swift_module.swift.response` -- almost entirely the history of these files consists of updates due to unrelated changes elsewhere in the codebase.

**Comment by lattner** [comment]
@gribozavr I'm not sure why avoiding cl::opt is important.  For point of comparison, clang uses its own complex driver, but clang-format uses cl::opt.

The reason to use a driver correlates to the complexity of the command line options, cl::opt seems appropriate for this tool.

File: tools/swift-format/swift-format.cpp [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    CodeFormatOptions FormatOptions;
+
+    class FormatterDocument {
+      SourceManager SM;
+      unsigned BufferID;
+      CompilerInvocation CompInv;
+      std::unique_ptr<ParserUnit> Parser;
+
+      class FormatterDiagConsumer: public swift::DiagnosticConsumer {
+
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Unnecessary whitespace

File: tools/swift-format/swift-format.cpp [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+        void handleDiagnostic(SourceManager &SM, SourceLoc Loc,
+                              DiagnosticKind Kind, StringRef Text,
+                              const DiagnosticInfo &Info) override {
+          llvm::errs() << "Parse error: " << Text << "\n";
+        }
+
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Unnecessary whitespace

File: tools/swift-format/swift-format.cpp [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    public:
+      FormatterDocument(std::unique_ptr<llvm::MemoryBuffer> Buffer) {
+        updateCode(std::move(Buffer));
+      }
+
+      void updateCode( std::unique_ptr<llvm::MemoryBuffer> buff ) {
+        BufferID = SM.addNewSourceBuffer(std::move(buff));
+
+        Parser.reset(
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Did clang-format really format this this way?

File: tools/swift-format/swift-format.cpp [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+                     new ParserUnit(SM, BufferID,
+                                    CompInv.getLangOptions(),
+                                    CompInv.getModuleName())
+                     );
+
+        Parser->getDiagnosticEngine().addConsumer(DiagConsumer);
+        auto &P = Parser->getParser();
+
+        bool Done = false;
+        while (!Done) {
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: This could probably be written using a `for` loop:

    for (bool Done = false; !Done; Done = P.Tok.is(tok::eof))
      P.parseTopLevel();

File: tools/swift-format/swift-format.cpp [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+      std::pair<unsigned, unsigned>
+      getLineAndColumn(unsigned ByteOffset) const {
+        auto &Buf = memBuffer();
+        if (ByteOffset > Buf.getBufferSize())
+          return std::make_pair(0, 0);
+
+        SMLoc Loc = SMLoc::getFromPointer(Buf.getBufferStart() + ByteOffset);
+        return SM.getLLVMSourceMgr().getLineAndColumn(Loc, BufferID);
+      }
+
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Unnecessary whitespace.

File: tools/swift-format/swift-format.cpp [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    // Parses <start line>:<end line> input to a pair of line numbers.
+    // Returns true on error.
+    static bool parseLineRange(StringRef Input, unsigned &FromLine,
+                               unsigned &ToLine) {
+      std::pair<StringRef, StringRef> LineRange = Input.split(':');
+      return LineRange.first.getAsInteger(0, FromLine) || LineRange.second.getAsInteger(0, ToLine);
+    }
+
+    static void outputReplacementXML(StringRef Text) {
+      // FIXME: When we sort includes, we need to make sure the stream is correct
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Left-overs from clang-format?  swift doesnt have includes ... I dont think that imports get sorted though.

File: tools/swift-format/swift-format.cpp [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        return true;
+      }
+      std::unique_ptr<llvm::MemoryBuffer> Code = std::move(CodeOrErr.get());
+      if (Code->getBufferSize() == 0)
+        return false; // Empty files are formatted correctly.
+
+      FormatterDocument Doc(std::move(Code));
+
+      if (!Offsets.empty() || !Lengths.empty()) {
+        if (Offsets.size() != Lengths.size()) {
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Cant this check be hoised before the construction of the memory buffer?

File: tools/swift-format/swift-format.cpp [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+            return true;
+          }
+          std::ostringstream s;
+          s << FromLine << ":" << ToLine;
+          LineRanges.push_back(s.str());
+        }
+      }
+
+      if (LineRanges.empty())
+        LineRanges.push_back("1:999999");
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Unfortunate; do we not have a constant for the end position?

File: tools/swift-format/swift-format.cpp [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+        }
+      }
+
+      if (OutputXML) {
+        llvm::outs() << "<?xml version='1.0'?>\n<replacements>\n";
+        outputReplacementsXML(Replaces);
+        llvm::outs() << "</replacements>\n";
+      } else {
+        if (Inplace) {
+          if (FileName == "-") {
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: Cant this check be hoist out before the memory buffer is constructed?

File: tools/swift-format/swift-format.cpp [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+            llvm::errs() << "error: cannot use -i when reading from stdin.\n";
+            return true;
+          }
+          else {
```
Comment (compnerd) [inline tools/swift-format/swift-format.cpp]: This should be joined with the previous line.

**Comment by danielmartin** [comment]
Thanks for your input. I'm implementing the suggested changes and I think I'll start modeling this as a driver tool, not as a standalone executable. I'm curious about any advantages of implementing it as a standalone binary, ie. why was clang-format implemented that way? I suppose it's because clang-format was thought to be downloaded separately from the Clang standard distribution. Am I right? /cc @lattner @gribozavr

**Comment by compnerd** [comment]
I personally think that a standalone binary is much nicer.  It allows for integration into various other tools, e.g. vim.  By having the tool be completely standalone, it doesnt require that I invoke it the same was as the driver.  All of the decisions about the code generation do not matter, and making it similar to the driver there makes it more complicated for no reason.

**Comment by gribozavr** [comment]
@compnerd It will have its own symlink, so there is no difference for the user, apart from the download size.

**Comment by lattner** [comment]
I agree with @gribozavr on this.  Assuming the user experience is the same, we should optimize compiler developers too :-)

**Comment by compnerd** [comment]
Yeah, as long as it can still be invoked the same, I think that a smaller binary is preferable.

**Comment by danielmartin** [comment]
@gribozavr @lattner I've re-implemented the tool as a compiler driver, and reduced the length of the sample program used for unit tests.

File: include/swift/Option/Options.h [hunk 11]
```diff
@@ -32,7 +32,8 @@ namespace options {
     NoBatchOption = (1 << 7),
     DoesNotAffectIncrementalBuild = (1 << 8),
     AutolinkExtractOption = (1 << 9),
-    ModuleWrapOption = (1 << 10)
+    ModuleWrapOption = (1 << 10),
+    SwiftFormatOption = (1 << 11)
```
Comment (compnerd) [inline include/swift/Option/Options.h]: Please add a trailing ,.  We are using C++11, so that should be fine.

File: include/swift/Option/Options.td [hunk 12]
```diff
@@ -310,6 +313,31 @@ def AssertConfig : Separate<["-"], "assert-config">,
   HelpText<"Specify the assert_configuration replacement. "
            "Possible values are Debug, Release, Unchecked, DisableReplacement.">;
 
+// Code formatting options
+
+def code_formatting_Group : OptionGroup<"<code formatting options>">;
+
+def use_tabs : Flag<["-"], "usetabs">, Group<code_formatting_Group>,
```
Comment (compnerd) [inline include/swift/Option/Options.td]: `use_tabs` would make be expect that the option is spelt `use-tabs`.

File: include/swift/Option/Options.td [hunk 13]
```diff
@@ -310,6 +313,31 @@ def AssertConfig : Separate<["-"], "assert-config">,
...
+  Flags<[SwiftFormatOption]>,
+  HelpText<"Use tabs for indentation.">;
+
+def format_inplace : Flag<["-"], "inplace">, Group<code_formatting_Group>,
```
Comment (compnerd) [inline include/swift/Option/Options.td]: `format_inplace` makes be think that the option should be spelt `format-inlace`.

File: lib/Driver/Driver.cpp [hunk 14]
```diff
@@ -2009,6 +2010,7 @@ void Driver::printHelp(bool ShowHidden) const {
     break;
   case DriverKind::Batch:
   case DriverKind::AutolinkExtract:
+  case DriverKind::SwiftFormat:
     ExcludedFlagsBitmask |= options::NoBatchOption;
```
Comment (compnerd) [inline lib/Driver/Driver.cpp]: It may be interesting to have this available as a batch mode operation, where you could format using the same invocation as a compile.

File: tools/driver/swift_format_main.cpp [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    }
+  }
+
+  std::pair<LineRange, std::string> reformat(LineRange Range,
+                                             CodeFormatOptions Options) {
+    return ::reformat(Range,
+                      Options,
+                      SM,
+                      Parser->getSourceFile());
+  }
```
Comment (compnerd) [inline tools/driver/swift_format_main.cpp]: clang-format this block?

File: tools/driver/swift_format_main.cpp [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+
+class SwiftFormatInvocation {
+private:
+  std::string MainExecutablePath;
+  std::string OutputFilename = "-";
+  std::vector<std::string> InputFilenames;
+  bool UseTabs = false;
+  bool InPlace = false;
+  unsigned TabWidth = 4;
+  unsigned IndentWidth = 4;
```
Comment (compnerd) [inline tools/driver/swift_format_main.cpp]: Why default to 4 when it seems that the default swift style is 2?

File: tools/driver/swift_format_main.cpp [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+    return 1;
+
+  std::vector<std::string> InputFiles = Invocation.getInputFilenames();
+  unsigned NumInputFiles = InputFiles.size();
+  if (NumInputFiles == 0)
+    // Read source code from standard input.
+    Invocation.format("-", Diags);
+  else if (NumInputFiles == 1)
+    Invocation.format(InputFiles[0], Diags);
+  else {
```
Comment (compnerd) [inline tools/driver/swift_format_main.cpp]: Please use braces around the if/else if/else here ... it tends to make it less readable if some of the blocks are braced and some are not.

File: tools/driver/swift_format_main.cpp (hunk 16 above)
Comment (danielmartin) [inline tools/driver/swift_format_main.cpp]: I've taken the defaults from the `CodeFormatOptions` struct in `Formatting.h`. In a future PR I plan to extract that into a "style" structure so that it is not duplicated across the codebase.

File: include/swift/Option/Options.h (hunk 11 above)
Comment (danielmartin) [inline include/swift/Option/Options.h]: Fixed: b41bdf027c5c8de53d8bfbb5f09eb29fe2aad77e

File: include/swift/Option/Options.td (hunk 12 above)
Comment (danielmartin) [inline include/swift/Option/Options.td]: Fixed: 8c5f74c3d2473cfb2fc163cda41d92085b18a9c1

File: include/swift/Option/Options.td (hunk 13 above)
Comment (danielmartin) [inline include/swift/Option/Options.td]: Fixed: 8c5f74c3d2473cfb2fc163cda41d92085b18a9c1

File: tools/driver/swift_format_main.cpp (hunk 15 above)
Comment (danielmartin) [inline tools/driver/swift_format_main.cpp]: Fixed: 6037125de81820e876a6772b7b666b8372ae4df8

File: tools/driver/swift_format_main.cpp (hunk 17 above)
Comment (danielmartin) [inline tools/driver/swift_format_main.cpp]: Fixed: 6037125de81820e876a6772b7b666b8372ae4df8

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
I think that we should squash the changes into a single commit.  I think that this is a reasonable place to start to merge this and then work on it iteratively.  Objections from @gribozavr or @lattner?

**Comment by gribozavr** [comment]
@compnerd No objections from me.

**Comment by danielmartin** [comment]
I've just squashed the patch into a single commit.

**Comment by compnerd** [comment]
@swift-ci please smoke test and merge

**Comment by gottesmm** [comment]
Failed due to some sort of unrelated exception on the bot. Merging!
